File: Python/6py 3/rainfall.py
"""
|--------------------------------------------------------------------------
| Rainfall
|--------------------------------------------------------------------------
|
"""

def parse_data(data):
   return {line.split(':')[0].strip(): [float(record.split()[1]) for record in line.split(':')[1].split(',')] for line in data.split('\n')}
def mean(town, data):
   city_data = parse_data(data)
   if town not in city_data: return -1.0
   rainfall = city_data[town]
   return sum(rainfall) / len(rainfall)
def variance(town, data):
   city_data = parse_data(data)
   if town not in city_data: return -1.0
   rainfall = city_data[town]
   mean_value = sum(rainfall) / len(rainfall)
   return sum((x - mean_value) ** 2 for x in rainfall) / len(rainfall)


import numpy as np
def split_data(town,strng):
   city_data = dict(city.split(":") for city in strng.splitlines())
   return dict(v.split() for v in [m for m in city_data[town].split(',')]) if town in city_data else 0
def mean(town, strng):
   d = split_data(town, strng)
   return np.average(np.array([float(x) for x in d.values()])) if d else -1
def variance(town, strng):
   d = split_data(town, strng)
   return np.var(np.array([float(x) for x in d.values()])) if d else -1

File: JavaScript/6js 2/rainfall.js
/*
|--------------------------------------------------------------------------
| Rainfall
|--------------------------------------------------------------------------
|
*/

function parseData(data) {
   return data.split("\n").reduce((acc, line) => {
      let [city, records] = line.split(":");
      acc[city.trim()] = records
         .split(",")
         .map((record) => parseFloat(record.trim().split(" ")[1]));
      return acc;
   }, {});
}
function mean(town, data) {
   let cityData = parseData(data);
   if (!(town in cityData)) return -1.0;
   let rainfall = cityData[town];
   return rainfall.reduce((sum, value) => sum + value, 0) / rainfall.length;
}
function variance(town, data) {
   let cityData = parseData(data);
   if (!(town in cityData)) return -1.0;
   let rainfall = cityData[town];
   let meanValue = mean(town, data);
   return (
      rainfall.reduce((sum, value) => sum + Math.pow(value - meanValue, 2), 0) / rainfall.length
   );
}


function parse(t, s) {
   var d = s
      .split("\n")
      .map((x) => x.split(":"))
      .filter((x) => x[0] == t);
   return d.length ? d[0][1].split(",").map((x) => +x.split(" ")[1]) : undefined;
}
function mean(t, s) {
   var d = parse(t, s);
   return d ? d.reduce((p, c) => p + c, 0) / d.length : -1;
}
function variance(t, s) {
   var d = parse(t, s);
   var m = d ? d.reduce((p, c) => p + c, 0) / d.length : -1;
   return m > -1 ? d.map((x) => (x - m) ** 2).reduce((p, c) => p + c, 0) / d.length : m;
}